algorithms/kernel/dtrees/forest/classification/oneapi/df_classification_train_hist_oneapi_impl.i

```diff
@@ -0,0 +1,933 @@
+
+    auto & kernel = kernelComputePartialHistograms;
+
+    {
+        KernelArguments args(11);
+        args.set(0, data, AccessModeIds::read);
+        args.set(1, treeOrder, AccessModeIds::read);
+        args.set(2, nodeList, AccessModeIds::read);
+        args.set(3, nodeIndices, AccessModeIds::read);
+        args.set(4, (int32_t)nodeIndicesOffset);
+        args.set(5, selectedFeatures, AccessModeIds::read);
+        args.set(6, response, AccessModeIds::read);
+        args.set(7, binOffsets, AccessModeIds::read);
+        args.set(8, (int32_t)nMaxBinsAmongFtrs); // max num of bins among all ftrs
+        args.set(9, (int32_t)nFeatures);
+        args.set(10, partialHistograms, AccessModeIds::write);
+
+        size_t localSize = nSelectedFeatures < _maxLocalSize ? nSelectedFeatures : _maxLocalSize;
+
+        KernelRange local_range(1, localSize, 1);
+        KernelRange global_range(nPartialHistograms, localSize, nNodes);
+
+        KernelNDRange range(3);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+
+template <typename algorithmFPType>
+services::Status ClassificationTrainBatchKernelOneAPI<algorithmFPType, hist>::reducePartialHistograms(
+    UniversalBuffer & partialHistograms, UniversalBuffer & histograms, size_t nPartialHistograms, size_t nNodes, size_t nSelectedFeatures,
+    size_t nMaxBinsAmongFtrs, size_t reduceLocalSize)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(compute.reducePartialHistograms);
+
+    services::Status status;
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+
+    auto & kernel = kernelReducePartialHistograms;
+
+    {
+        KernelArguments args(5);
+        args.set(0, partialHistograms, AccessModeIds::read);
+        args.set(1, histograms, AccessModeIds::write);
+        args.set(2, (int32_t)nPartialHistograms);
+        args.set(3, (int32_t)nSelectedFeatures);
+        args.set(4, (int32_t)nMaxBinsAmongFtrs); // max num of bins among all ftrs
+
+        KernelRange local_range(1, reduceLocalSize, 1);
+        KernelRange global_range(nMaxBinsAmongFtrs * nSelectedFeatures, reduceLocalSize, nNodes);
+
+        KernelNDRange range(3);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+
+template <CpuType cpu>
+static void shuffle(void * state, size_t n, int * dst)
+{
+    RNGs<int, cpu> rng;
+    int idx[2];
+
+    for (size_t i = 0; i < n; ++i)
+    {
+        rng.uniform(2, idx, state, 0, n);
+        daal::services::internal::swap<cpu, int>(dst[idx[0]], dst[idx[1]]);
```

Comment:
This code does not actually produce equiprobability of all possible permutations.
Right way to do that is the following:
```
    for (size_t i = 0; i < n; ++i)
    {
        rng.uniform(1, idx, state, 0, i + 1);
        daal::services::internal::swap<cpu, int>(dst[idx[0]], dst[i]);
    }
```

Reply:
thank you, for now it is synchronized with CPU part, will change it during migration on GPU generators